cpp/inc/bond/core/capped_allocator.h

```diff
@@ -0,0 +1,313 @@
+        {}
+
+        pointer allocate(size_type n)
+        {
+            const auto size = n * sizeof(value_type);
+
+            if (_count.get().try_add(size))
+            {
+                try
+                {
+                    return traits::allocate(get_allocator(), n);
+                }
+                catch (...)
+                {
+                    _count.get().subtract(size);
+                    throw;
+                }
+            }
+
+            throw std::bad_alloc{};
+        }
+
+        pointer allocate(size_type n, const_void_pointer hint)
+        {
+            const auto size = n * sizeof(value_type);
+
+            if (_count.get().try_add(size))
+            {
+                try
+                {
+                    return traits::allocate(get_allocator(), n, hint);
+                }
+                catch (...)
+                {
+                    _count.get().subtract(size);
+                    throw;
+                }
+            }
+
+            throw std::bad_alloc{};
+        }
+
+        void deallocate(pointer ptr, size_type n)
+        {
+            traits::deallocate(get_allocator(), ptr, n);
+
+            if (_subtract_on_deallocate)
+            {
+                _count.get().subtract(n * sizeof(value_type));
+            }
+        }
+
+        template <typename T, typename... Args>
+        void construct(T* ptr, Args&&... args)
+        {
+            traits::construct(get_allocator(), ptr, std::forward<Args>(args)...);
+        }
+
+        template <typename T>
+        void destroy(T* ptr)
+        {
+            traits::destroy(get_allocator(), ptr);
+        }
+
+        size_type max_size() const BOND_NOEXCEPT
+        {
+            return (std::min)(
+                _count.get().max_value() / sizeof(value_type),
+                traits::max_size(get_allocator()));
+        }
+
+        capped_allocator select_on_container_copy_construction()
+        {
+            return capped_allocator{
+                _count,
+                traits::select_on_container_copy_construction(get_allocator()),
+                _subtract_on_deallocate };
+        }
+
+        const Alloc& get_allocator() const BOND_NOEXCEPT
+        {
+            return holder::get();
+        }
+
+        Alloc& get_allocator() BOND_NOEXCEPT
+        {
+            return holder::get();
+        }
+
+        const Counter& get_counter() const BOND_NOEXCEPT
+        {
+            return _count.get();
+        }
+
+        Counter& get_counter() BOND_NOEXCEPT
+        {
+            return _count.get();
+        }
+
+    private:
+        template <typename OtherAlloc, typename OtherCounter>
+        friend class capped_allocator;
+
+        detail::value_or_reference<Counter> _count;
+        bool _subtract_on_deallocate;
+    };
+
+
+    template <typename Alloc, typename Counter>
+    inline bool operator==(
```

Comment:
The comparison is required by [allocator concept](http://en.cppreference.com/w/cpp/concept/Allocator) and is used by STL containers. The counter comparison is skipped intentionally, because the underlying allocator is the same and it is OK to use the optimal code path (e.g. move the whole [std::vector](http://en.cppreference.com/w/cpp/container/vector/vector) instead of doing element-wise move).